perf/light.cc

```diff
@@ -0,0 +1,651 @@
+		Erase(data.tuples.begin(), data.tuples.end());
+		data.shuffle();
+		state.ResumeTiming();
+
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_insert(&ht, v.hash, v.raw));
+			processed++;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+BENCHMARK_REGISTER_F(LightFixture, InsertAfterErase)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+// Fill in hash table, then erase half elements and process lookups
+// (only lookup time is measured).
+BENCHMARK_DEFINE_F(LightFixture, FindAfterErase)
+(benchmark::State& state)
+{
+	std::size_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		// Double the data size so that after shrink it'll be of the
+		// same size as in other benches.
+		std::size_t data_size = state.range(0) * 2;
+		TupleHolder data(data_size);
+		Fill(data.tuples.begin(), data.tuples.end());
+		Erase(data.tuples.begin(), std::next(data.tuples.begin(), data_size / 2));
+		data.shuffle();
+		state.ResumeTiming();
+
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_find(&ht, v.hash, v.raw));
+			processed++;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+BENCHMARK_REGISTER_F(LightFixture, FindAfterErase)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+// Random insert and random lookup. On each iteration it processes insert and
+// after that - lookup (half of lookups - misses).
+BENCHMARK_DEFINE_F(LightFixture, InsertAndFind)
+(benchmark::State& state)
+{
+	std::size_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		state.ResumeTiming();
+
+		auto lookup_iter = data.tuples.rbegin();
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_insert(&ht, v.hash, v.raw));
+			benchmark::DoNotOptimize(light_find(&ht,
+				lookup_iter->hash, lookup_iter->raw));
+			lookup_iter++;
+			processed += 2;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+BENCHMARK_REGISTER_F(LightFixture, InsertAndFind)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+// Delete random values till the hashtable is empty.
+BENCHMARK_DEFINE_F(LightFixture, DeleteRandValue)
+(benchmark::State& state)
+{
+	std::size_t processed = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		Fill(data.tuples.begin(), data.tuples.end());
+		data.shuffle();
+		state.ResumeTiming();
+
+		for (const auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_delete_value(&ht, v.hash, v.raw));
+			processed++;
+		}
+	}
+	state.SetItemsProcessed(processed);
+}
+
+BENCHMARK_REGISTER_F(LightFixture, DeleteRandValue)->
+	RangeMultiplier(TUPLE_COUNT_MULTIPLIER)->
+		Range(TUPLE_COUNT_MIN, TUPLE_COUNT_MAX);
+
+////////////////////////////// Unordered set Fixture ///////////////////////////////////////////////////////////////////
+
+struct TupleEqual
```

Comment:
It seems that this class should be places right near TupleHash.

Reply:
Moved